File: MyMiniProject/angular-dotnet-finance/backend/Controllers/TransactionsController.cs
using System.Security.Claims;
using FinanceApi.Data;
using FinanceApi.Models;
using FinanceApi.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace FinanceApi.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class TransactionsController(AppDbContext db, SlipStorage slips) : ControllerBase
{
    private int CurrentUserId => int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier)!);
    private bool IsAdmin => User.IsInRole("admin");

    [HttpGet]
    public async Task<ActionResult<List<TransactionDto>>> GetAll(
        [FromQuery] string? type,
        [FromQuery] string? category,
        [FromQuery] int? userId)
    {
        var q = db.Transactions.Include(t => t.User).AsQueryable();

        if (!IsAdmin)
            q = q.Where(t => t.UserId == CurrentUserId);
        else if (userId is not null)
            q = q.Where(t => t.UserId == userId);

        if (!string.IsNullOrWhiteSpace(type))
            q = q.Where(t => t.Type == type);

        if (!string.IsNullOrWhiteSpace(category))
            q = q.Where(t => t.Category.Contains(category));

        var list = await q.OrderByDescending(t => t.OccurredAt).ThenByDescending(t => t.Id).ToListAsync();
        return Ok(list.Select(ToDto).ToList());
    }

    [HttpGet("{id:int}")]
    public async Task<ActionResult<TransactionDto>> Get(int id)
    {
        var t = await db.Transactions.Include(x => x.User).FirstOrDefaultAsync(x => x.Id == id);
        if (t is null) return NotFound();
        if (!IsAdmin && t.UserId != CurrentUserId) return Forbid();
        return Ok(ToDto(t));
    }

    [HttpPost]
    [RequestSizeLimit(6 * 1024 * 1024)]
    public async Task<ActionResult<TransactionDto>> Create([FromForm] TransactionForm form)
    {
        var validation = Validate(form);
        if (validation is not null) return validation;

        var ownerId = IsAdmin && form.UserId is > 0 ? form.UserId.Value : CurrentUserId;
        var owner = await db.Users.FindAsync(ownerId);
        if (owner is null)
            return BadRequest(new { message = "ไม่พบผู้ใช้" });

        var t = new Transaction
        {
            UserId = ownerId,
            Type = form.Type!.Trim().ToLowerInvariant(),
            Category = form.Category!.Trim(),
            Title = form.Title!.Trim(),
            Note = form.Note?.Trim(),
            Amount = form.Amount,
            OccurredAt = form.OccurredAt ?? DateTime.UtcNow
        };

        db.Transactions.Add(t);
        await db.SaveChangesAsync();

        try
        {
            if (form.Slip is not null)
            {
                t.SlipPath = await slips.SaveAsync(form.Slip, t.Id, owner.Username);
                await db.SaveChangesAsync();
            }
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }

        await db.Entry(t).Reference(x => x.User).LoadAsync();
        return CreatedAtAction(nameof(Get), new { id = t.Id }, ToDto(t));
    }

    [HttpPut("{id:int}")]
    [RequestSizeLimit(6 * 1024 * 1024)]
    public async Task<ActionResult<TransactionDto>> Update(int id, [FromForm] TransactionForm form)
    {
        var t = await db.Transactions.Include(x => x.User).FirstOrDefaultAsync(x => x.Id == id);
        if (t is null) return NotFound();
        if (!IsAdmin && t.UserId != CurrentUserId) return Forbid();

        var validation = Validate(form);
        if (validation is not null) return validation;

        t.Type = form.Type!.Trim().ToLowerInvariant();
        t.Category = form.Category!.Trim();
        t.Title = form.Title!.Trim();
        t.Note = form.Note?.Trim();
        t.Amount = form.Amount;
        if (form.OccurredAt.HasValue) t.OccurredAt = form.OccurredAt.Value;

        if (IsAdmin && form.UserId is > 0)
            t.UserId = form.UserId.Value;

        var owner = await db.Users.FindAsync(t.UserId);
        if (owner is null)
            return BadRequest(new { message = "ไม่พบผู้ใช้" });

        try
        {
            if (form.ClearSlip)
            {
                slips.DeleteIfExists(t.SlipPath);
                t.SlipPath = null;
            }

            if (form.Slip is not null)
            {
                slips.DeleteIfExists(t.SlipPath);
                t.SlipPath = await slips.SaveAsync(form.Slip, t.Id, owner.Username);
            }
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { message = ex.Message });
        }

        await db.SaveChangesAsync();
        await db.Entry(t).Reference(x => x.User).LoadAsync();
        return Ok(ToDto(t));
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        var t = await db.Transactions.FindAsync(id);
        if (t is null) return NotFound();
        if (!IsAdmin && t.UserId != CurrentUserId) return Forbid();

        slips.DeleteIfExists(t.SlipPath);
        db.Transactions.Remove(t);
        await db.SaveChangesAsync();
        return NoContent();
    }

    private ActionResult? Validate(TransactionForm form)
    {
        if (string.IsNullOrWhiteSpace(form.Title))
            return BadRequest(new { message = "กรุณาระบุหัวข้อ" });
        if (string.IsNullOrWhiteSpace(form.Category))
            return BadRequest(new { message = "กรุณาระบุหมวดหมู่" });
        if (form.Type is not ("income" or "expense"))
            return BadRequest(new { message = "Type ต้องเป็น income หรือ expense" });
        if (form.Amount <= 0)
            return BadRequest(new { message = "จำนวนเงินต้องมากกว่า 0" });
        return null;
    }

    private static TransactionDto ToDto(Transaction t) => new(
        t.Id,
        t.UserId,
        t.User?.FullName,
        t.Type,
        t.Category,
        t.Title,
        t.Note,
        t.Amount,
        t.OccurredAt,
        t.SlipPath,
        t.CreatedAt);
}

public class TransactionForm
{
    public int? UserId { get; set; }
    public string? Type { get; set; }
    public string? Category { get; set; }
    public string? Title { get; set; }
    public string? Note { get; set; }
    public decimal Amount { get; set; }
    public DateTime? OccurredAt { get; set; }
    public IFormFile? Slip { get; set; }
    public bool ClearSlip { get; set; }
}

File: MyMiniProject/laravel-finance/app/Http/Controllers/Api/TransactionsController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Transaction;
use App\Models\User;
use App\Services\SlipStorage;
use Illuminate\Http\Request;

class TransactionsController extends Controller
{
    public function __construct(private SlipStorage $slips) {}

    public function index(Request $request)
    {
        /** @var User $auth */
        $auth = $request->attributes->get('auth_user');
        $q = Transaction::with('user')->orderByDesc('occurred_at')->orderByDesc('id');

        if (! $auth->isAdmin()) {
            $q->where('user_id', $auth->id);
        } elseif ($request->filled('userId')) {
            $q->where('user_id', (int) $request->query('userId'));
        }

        if ($request->filled('type')) {
            $q->where('type', $request->query('type'));
        }
        if ($request->filled('category')) {
            $q->where('category', 'like', '%'.$request->query('category').'%');
        }

        return response()->json($q->get()->map->toApiArray()->values());
    }

    public function show(Request $request, int $id)
    {
        /** @var User $auth */
        $auth = $request->attributes->get('auth_user');
        $t = Transaction::with('user')->find($id);
        if (! $t) {
            return response()->json(['message' => 'Not found'], 404);
        }
        if (! $auth->isAdmin() && $t->user_id !== $auth->id) {
            return response()->json(['message' => 'Forbidden'], 403);
        }

        return response()->json($t->toApiArray());
    }

    public function store(Request $request)
    {
        return $this->upsert($request, null);
    }

    public function update(Request $request, int $id)
    {
        return $this->upsert($request, $id);
    }

    public function destroy(Request $request, int $id)
    {
        /** @var User $auth */
        $auth = $request->attributes->get('auth_user');
        $t = Transaction::find($id);
        if (! $t) {
            return response()->json(['message' => 'Not found'], 404);
        }
        if (! $auth->isAdmin() && $t->user_id !== $auth->id) {
            return response()->json(['message' => 'Forbidden'], 403);
        }

        $this->slips->deleteIfExists($t->slip_path);
        $t->delete();

        return response()->noContent();
    }

    private function upsert(Request $request, ?int $id)
    {
        /** @var User $auth */
        $auth = $request->attributes->get('auth_user');

        $title = trim((string) $request->input('title', ''));
        $category = trim((string) $request->input('category', ''));
        $type = strtolower(trim((string) $request->input('type', '')));
        $amount = (float) $request->input('amount', 0);

        if ($title === '') {
            return response()->json(['message' => 'กรุณาระบุหัวข้อ'], 400);
        }
        if ($category === '') {
            return response()->json(['message' => 'กรุณาระบุหมวดหมู่'], 400);
        }
        if (! in_array($type, ['income', 'expense'], true)) {
            return response()->json(['message' => 'Type ต้องเป็น income หรือ expense'], 400);
        }
        if ($amount <= 0) {
            return response()->json(['message' => 'จำนวนเงินต้องมากกว่า 0'], 400);
        }

        $ownerId = $auth->id;
        if ($auth->isAdmin() && (int) $request->input('userId') > 0) {
            $ownerId = (int) $request->input('userId');
        }

        $owner = User::find($ownerId);
        if (! $owner) {
            return response()->json(['message' => 'ไม่พบผู้ใช้'], 400);
        }

        if ($id) {
            $t = Transaction::with('user')->find($id);
            if (! $t) {
                return response()->json(['message' => 'Not found'], 404);
            }
            if (! $auth->isAdmin() && $t->user_id !== $auth->id) {
                return response()->json(['message' => 'Forbidden'], 403);
            }
        } else {
            $t = new Transaction;
        }

        $t->user_id = $ownerId;
        $t->type = $type;
        $t->category = $category;
        $t->title = $title;
        $t->note = trim((string) $request->input('note', '')) ?: null;
        $t->amount = $amount;
        if ($request->filled('occurredAt')) {
            $t->occurred_at = $request->input('occurredAt');
        } elseif (! $id) {
            $t->occurred_at = now();
        }

        $t->save();

        try {
            $clear = filter_var($request->input('clearSlip'), FILTER_VALIDATE_BOOLEAN);
            if ($clear) {
                $this->slips->deleteIfExists($t->slip_path);
                $t->slip_path = null;
            }

            if ($request->hasFile('slip')) {
                $this->slips->deleteIfExists($t->slip_path);
                $t->slip_path = $this->slips->save($request->file('slip'), $t->id, $owner->username);
            }
            $t->save();
        } catch (\InvalidArgumentException $e) {
            return response()->json(['message' => $e->getMessage()], 400);
        }

        $t->load('user');

        return response()->json($t->toApiArray(), $id ? 200 : 201);
    }
}
